<!-- 
This document is autogenerated by markdownDocs() in CommandList.java. Do not edit it here. Edit source code CommandList.markdownDocs() then run tests in CommandListTest.updateMarkdownFile() to recreate this file.
-->

# Help to inline commands

This is the documentation for the indvidual commands. Note that the help documented here can be invoked also at the command prompt with `command -h`, for example to get the help for `ylim`:
```
ylim -h
```
Parameters in square brakets are optional and the default argument is indicated by the `=` sign. The syntax `...` indicate that the argument can be repeated multiple times. For example
```
ylim min max [track_regex = .*]...
```
Means that `ylim` takes two mandatory arguments, `min` and `max`. The optional argument, `track_regex`, defaults to `.*` and can be repated multiple times.


- [Navigation](#navigation)
  - [f](#f)
  - [b](#b)
  - [ff](#ff)
  - [bb](#bb)
  - [goto](#goto)
  - [zi](#zi)
  - [zo](#zo)
  - [INT](#int)
  - [+](#+)
  - [-](#-)
  - [p](#p)
  - [n](#n)
  - [next](#next)
- [Find](#find)
  - [find](#find)
  - [seqRegex](#seqregex)
- [Display](#display)
  - [grep](#grep)
  - [squash](#squash)
  - [merge](#merge)
  - [gap](#gap)
  - [gffNameAttr](#gffnameattr)
  - [trackHeight](#trackheight)
  - [ylim](#ylim)
  - [colorTrack](#colortrack)
  - [hideTitle](#hidetitle)
  - [dataCol](#datacol)
  - [print](#print)
- [Alignments](#alignments)
  - [rpm](#rpm)
  - [samtools](#samtools)
  - [BSseq](#bsseq)
- [General](#general)
  - [setGenome](#setgenome)
  - [showGenome](#showgenome)
  - [infoTracks](#infotracks)
  - [addTracks](#addtracks)
  - [dropTracks](#droptracks)
  - [orderTracks](#ordertracks)
  - [history](#history)
  - [cmdHistory](#cmdhistory)
  - [save](#save)
  - [q](#q)
  - [h](#h)


## Navigation

### f

**Usage: f [NUM=0.1]**

Move forward NUM times the size of the current window, 1/10 by default. 

### b

**Usage: b [NUM=0.1]**

Move backward NUM times the size of the current window, 1/10 by default 

### ff

**Usage: ff**

Move forward by 1/2 of a window. A shortcut for `f 0.5` 

### bb

**Usage: bb**

Move backward by 1/2 of a window. A shortcut for `b 0.5` 

### goto

**Usage: goto chrom:[from]-[to]**

Go to region chrom:from-to or to chrom:from or to start of chrom.  The character ':' is a shortcut for `goto`. Examples:
```
goto chr8:1-1000  ## Go to interval 1-1000 on chr8
goto chr8:10      ## Go to position 10 on chr8
goto chr8         ## Go to start of chr8
```
Or the same
```
:chr8:1-1000 
:chr8:10 
:chr8
```

### zi

**Usage: zi [INT = 1]**

Zoom in INT times. Each zoom halves the window size.  To zoom quickly use INT= 5 or 10 e.g. `zi 10`

### zo

**Usage: zo [INT = 1]**

Zoom out INT times. Each zoom doubles the window size.  To zoom quickly use INT= 5 or 10 e.g. `zo 10`

### INT

**Usage: INT [INT]**

Go to position `INT` or to region `INT INT` on current chromosome.  Allowed is the hyphenated format  separating the two positions. If a list of integers is given, the first and last are taken as *from* and *to*. This is handy to copy and paste intervals from the ruler above the prompt. 
Examples:
```
10                   ## Will jump to position 10 
10 1000              ## Go to region 10-1000 
10-1000              ## Same as above
10 250 500 750 1000  ## Same as above again
```


### +

**Usage: + INT [k|m]**

Move forward by INT bases. Suffixes k (kilo) and M (mega) are expanded to x1000 and x1,000,000. Examples: `-2m` or `+10k` or `+10.5k` 

### -

**Usage: - INT [k|m]**

Move backwards by INT bases. Suffixes k (kilo) and M (mega) are expanded to x1000 and x1,000,000.
Examples: `-100` or `-10k` or `-10.5m` 

### p

**Usage: p**

Go to the previous visited position. Similar to the back and forward arrows of an Internet browser.

### n

**Usage: n**

Go to the next visited position. Similar to the back and forward arrows of an Internet browser.

### next

**Usage: next [-start] [track_id]**

Move to the next feature on track_id on *current* chromosome.  `next` centers the window on the found feature and zooms out. This is useful for quickly browsing through annotation files of genes or ChIP-Seq peaks in combination with read coverage tracks (bigwig, tdf, etc.). The `-start` flag sets the window right at the start of the feature, without centering and zooming out.

The `next` command does exactly that, it moves to the next feature. If there are no more features after the current position it doesn't rewind to the beginning (use `1` for that) and it doesn't move to another chromosome, use `goto chrom` for that.
 
If `track_id` is omitted, the first annotation track is used. If trackId is not a feature track (bed, gtf, etc) a more or less ugly warning is issued.

## Find

### find

**Usage: find [-all] regex [track_id]**

Find the first record in track_id containing regex. The regex search starts from the *end* of the current window (so the current window is not searched) and moves forward on the current chromosome. At the end  of the current chromosome move to the next chromosomes and then restart at  the start of the initial one. The search stops at the first match found. If track_id is omitted the first interval track found is searched. The `-all` flag will return the region conatining all the regex matches.
Examples:
find -all ACTB genes.gtf -> Find the region containing all the matches of ACTB, including e.g. LACTB
find 'ACTB gene' -> Find the first match of 'ACTB gene' (note single quotes)

### seqRegex

**Usage: seqRegex [-c] [regex]**

Find regex in reference sequence and show matches as and additional track.  By default matching is irrespective of case unless the flag `-c` is set. Useful to show restriction enzyme sites, transcription factor motifs, etc. If the regex is omitted the matching is disabled. Examples
```
seqRegex ACTG    -> Case insensitive, actg matched
seqRegex -c ACTG -> Case sensitive, will not match actg
seqRegex         -> Disable regex matching track
```
This command is ignored if the reference fasta file is missing.

## Display

### grep

**Usage: grep [-i = .*] [-e = ''] [track_regex = .*]...**

Similar to grep command, filter for features matching `-i` regex, hide those matching `-e` regex. Apply to tracks matched by track_regex. This command is useful to filter the annotation in GTF or BED files, for example:
```
grep -i RNA -e mRNA gtf gff
```
Will show the rows containing 'RNA' but will hide those containing 'mRNA', applies to tracks whose name matches 'gtf' or 'gff'.
With no arguments reset to default: `grep -i .* -e ^$ .*` which means show everything, hide nothing, apply to all tracks.

### squash

**Usage: squash [track_regex = .*]...**

Toggle the squashing of features with the same coordinates.  If set, features with the same start, end, and strand are squashed in a single one. The displayed feature is the first one found in the group of features with the same coordinates. Useful to compact GTF where e.g. CDS and exons have the same coordinates. Applies only to annotation tracks captured by track_regex

### merge

**Usage: merge [track_regex = .*]...**

Toggle the merging of overlapping features. If set, features with overalapping coordinates are merged in a single one. Merged features will not have strand and name information. Note that merging is done without considering strand information. Applies only to annotation tracks captured by the list of track_regex

### gap

**Usage: gap [track_regex = .*]...**

Toggle the switch to add a gap between features. Default is true. If gap is set, as per default, features which on screen do not have at least one space separating are moved to different lines so that it is clear where is the end of one feature  and the start of the next one. If gap is unset such features might appear as a single, continuous one instaed. 
Example with gap set:
```
||||||
      ||||||
```
With gap unset these two features look like:
```
||||||||||||
```
Gap unset is preferable when the interest is in knowing which regions are covered since it gives a more compact view and the distiction betwen adjacent features is not important.

### gffNameAttr

**Usage: gffNameAttr [attribute_name = NULL] [track_regex = .*]...**

For GTF/GFF tracks, choose the attribute to get the feature name from. Use attribute NULL to reset to default choice of attribute. Applies to all GFF/GTF tracks captured by the list of `track_regex`. Example, given the gtf feature
```
chr1 . CDS  10 99 . + 2 gene_id "PTGFRN"; transcript_id "NM_020440";
```
Use gene_name as feature name or transcript_id:
```
gffNameAttr gene_name genes.gtf .*gff
PTGFRN_CCCCCCCCC

gffNameAttr transcript_id genes.gtf .*gff
NM_020440_CCCCCC
```


### trackHeight

**Usage: trackHeight INT [track_regex = .*]...**

Set track height to INT lines of text for all tracks matching regexes.  Setting height to zero hides the track and skips the processing altogether. This is useful to speed up the browsing when large bam files are present. Use infoTrack to see which tracks are hidden. Example:
`trackHeight 5 aln.*bam gtf`

### ylim

**Usage: ylim min max [track_regex = .*]...**

Set the y-axis limit for all tracks matched by regexes. Use `na` to autoscale to min and/or max. This command applies only to tracks displaying quantitative data on y-axis (e.g. bigwig, tdf), the other tracks are unaffected.
Examples:
```
ylim 0 50      ## Set min= 0 and max= 50 in all tracks.
ylim 0 na      ## Set min to 0 and autoscale the max. Apply to all tracks
ylim na na tdf ## Autoscale min and max. Apply to all tracks matching 'tdf'
```


### colorTrack

**Usage: colorTrack color [track_regex = .*]...**

Set colour for tracks matched by regex.  Available colours: red, green, yellow, blue, magenta, cyan, grey, light_red, light_green, light_yellow, light_blue, light_magenta, light_cyan, light_grey, white, black, default. The 'default' colour reset to the system default colour. Colouring is rendered with ANSI codes 8/16. Example:
```
colorTrack light_blue ts.*gtf ts.*bam
```


### hideTitle

**Usage: hideTitle [track_regex = .*]...**

Toggle the display of the title line matched by track_regex. Use /hide_all/ and /show_all/ to hide all tracks or show all tracks instead of toggling their mode. Examples:
```
hideTitle            -> Toggle all tracks, same as hideTitle .*
hideTitle bam bed    -> Toggle all tracks matched by 'bam' or 'bed'
hideTitle /hide_all/ -> Hide all tracks
```

### dataCol

**Usage: dataCol [index = 4] [track_regex = .*]...**

Select data column for bedgraph tracks containing regex.  index: 1-based column index. This command applies only to tracks of type bedgraph.
 For example, use column 5 on tracks containing #1 and #3:
 ```
dataCol 5 #1 #3
```


### print

**Usage: print [-full] [track_regex = .*]...**

Toggle the printing of lines for the tracks matched by track_regex.  Useful to show exactly what features are present in the current window. Features are filtered in/out according to the `grep` command. Applies only to annotation tracks. Lines extending beyond the screen width are clipped for readability unless the  the flag `-full` is enabled.

## Alignments

### rpm

**Usage: rpm [track_regex = .*]**

Toggle read coverage from raw count to reads per million. 

### samtools

**Usage: samtools [-f INT=0] [-F INT=4] [-q INT=0] [track_re = .*] ...**

Apply samtools filters to alignment tracks captured by the list of track regexes. As *samtools view*, this command filters alignment records on the basis of the given flags:
-F: Filter out flags with these bits set. NB: 4 is always set.
-f: Require alignment to have these bits sets.
-q: Require alignments to have MAPQ >= than this. Examples:
```
samtools -q 10           -> Set mapq for all tracks. -f and -F reset to default
samtools -F 1024 foo bar -> Set -F for all track containing re foo or bar
samtools                 -> Reset all to default.
```

### BSseq

**Usage: BSseq [track_regex = .*]...**

Toggle bisulfite mode for read tracks matched by regex. In bisulfite mode, the characters M and m mark methylated bases (i.e. unconverted C to T) and U and u are used for unmethylated bases (i.e. C converted to T). Upper case is used for reads on  forward strand, small case for reverse. Ignored without reference fasta sequence.

## General

### setGenome

**Usage: setGenome fasta|bam|genome**

Set genome and reference sequence. The genome, i.e. the list of contig and names and sizes, can be extracted from the indexed fasta reference, from a bam file or from a genome identifier (e.g. hg19). If a fasta file is used also the reference sequence becomes available.

### showGenome

**Usage: showGenome**

Print the genome dictionary with a representation of chromosome sizes.  Example output:
```
showGenome
chrM  16571
chr1  249250621 ||||||||||||||||||||||||||||||
chr2  243199373 |||||||||||||||||||||||||||||
...
chr21 48129895  ||||||
chr22 51304566  ||||||
chrX  155270560 |||||||||||||||||||
chrY  59373566  |||||||
```


### infoTracks

**Usage: infoTracks**

Print the name of the current tracks along with file name and format.  Hidden tracks are marked by *.

### addTracks

**Usage: addTracks [file or URL]...**

Add tracks from local or remote files. 
Examples:
```
addTracks peaks.bed gene.gtf
addTracks http://remote/host/peaks.bed
```

### dropTracks

**Usage: dropTracks track_regex [track_regex]...**

Drop tracks matching any of the listed regexes. 
Examples:
```
dropTracks bam
```

### orderTracks

**Usage: orderTracks [track_regex]...**

Reorder tracks according to the list of regexes or sort by name. Not all the tracks need to be listed, the missing ones follow the listed ones in unchanged order. Without arguments sort track by tag name.
For example, given the track list: `[hela.bam#1, hela.bed#2, hek.bam#3, hek.bed#4]`:

```
orderTracks #2 #1   new order: [hela.bed#2, hela.bam#1, hek.bam#3, hek.bed#4]
orderTracks bam bed new order: [hela.bam#1, hek.bam#3, hela.bed#2, hek.bed#4]
orderTracks         sort by tag [hela.bam#1, hela.bed#2, hek.bam#3, hek.bed#4]
```

### history

**Usage: history**

Show the list of visited positions. 

### cmdHistory

**Usage: cmdHistory**

Show the list of executed commands. 

### save

**Usage: save [filename = chrom_start_end.txt']**

Save current screenshot to file in either text or png format. Default filename is generated from the current coordinates and the default format is txt. With filename .png save as png using current coordinates as filename. Use extension .png to save as png format. Colours are stripped from text output but they retainedin the png. The string `%r` in the file name is replaced with the current coordinates.Examples:
```
save mygene.txt ## Save to mygene.txt as text
save            ## Save to chrom_start-end.txt as text
save .png       ## Save to chrom_start-end.png as png
save mygene.%r.png ## Save to mygene.chr1_100-200.png as png
```


### q

**Usage: q**

Quit 

### h

**Usage: h**

Show this help. For help on specific commands use `command -h`, for example `ylim -h` 


